### PR TITLE
Graphics tests docs

### PR DESCRIPTION
**Pull request by pp-mo**
First cut at documenting new graphics test arrangements + workflow.

**NOT yet complete**

### CODE REVIEW

**Comment by ajdawson** [comment]
I haven't been following this closely but it seems like an interesting tactic for handling image tests across dependency versions and platforms. The one problem is that the hashes of 2 very similar images will be very different, meaning we need exact matches to one of our stored hashes. This seems OK as long as we can generate enough hashes, but I wonder how well this will scale in the future.

It might be valuable to consider using a different hashing algorithm, one where similar images generate hashes that are "close". Luckily such things exist already, called perceptual hashes. A quick search suggests Python implementations of [pHash](http://www.phash.org/docs/design.html) are available (e.g. https://github.com/JohannesBuchner/imagehash). This would allow storing hashes for major changes (e.g. when matplotlib changes something like default colormap) but minor variants around these could be matched by checking the distance between two hashes rather than having to add new reference hashes for each case. Perhaps worth considering.

**Comment by bjlittle** [comment]
Thanks @ajdawson!

Perceptual image hashing has been on our radar and already discussed. See [here](http://nbviewer.jupyter.org/gist/bjlittle/7a8adda9d4dd351bcdf1a9a82f5ca63e) for an initial investigation. It raises the interesting question of how "similar" is a result to an expected result, and is that definition of "similarity" acceptable to suit our testing needs. We thought that we'd first exercise the use of a zero tolerance cryptographic hash approach and assess the impact and practicalities of demanding exact image equivalence rather than "similar" images based on a hamming distance hash threshold.

It's certainly a very interesting problem space. With the new work flow pattern and framework that we now have in place for image testing, I think we're in a much healthier place i.e. it's an enabler that's got us pretty darn close to unpinning mpl within Travis from 1.3 to latest mpl 1.5+, which in turn releases us from the numpy 1.8 shackles, see #2124 . So I believe perceptual image hashing is still a viable option that we can chose to use and it's a question that I think we should consider answering sooner rather than later ...

**Comment by marqh** [comment]
> it's a question that I think we should consider answering sooner rather than later ...

indeed

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+There should be sufficient work-flow detail here to allow an iris developer to:
+    * understand the new check graphic test process
+    * understand the steps to take and tools to use to add a new graphic test
+    * understand the steps to take and tools to use to diagnose and fix an graphic test failure
+
+
+Basic workflow
+==============
+#. a graphics test result has changed, following changes in Iris or
+   dependencies, so a test is failing
```
Comment (corinnebosley) [inline docs/iris/src/developers_guide/graphics_tests.rst]: I like this point.  It's important to include a trigger or reason for having to do this.

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#. the hash value of the new result is added into the relevant set of 'valid
+   result hashes' in the image result database file,
+   ``tests/results/imagerepo.json``.
+#. the tests can now be re-run, and the 'new' result should now be accepted.
+#. a pull request is created to add this file into the test-images-scitools
+   repository
+#. a pull request is created in the Iris repository, including the change to
+   the image results database (``tests/results/imagerepo.json``) :
+   This pull request must contain a reference to the matching one in
+   test-images-scitools.
```
Comment (corinnebosley) [inline docs/iris/src/developers_guide/graphics_tests.rst]: Good stubs.  I agree that enumerating the steps will make it easier to add the necessary detail.

File: docs/iris/src/developers_guide/pulls.rst [hunk 3]
```diff
@@ -16,66 +16,78 @@ is merged.  Before submitting a pull request please consider this list.
...
+  * iris-test-data_ is a github project containing all the data to support the
+    tests.
+  * iris-sample-data_ is a github project containing all the data to support
+    the gallery and examples.
+  * test-images-scitools_ is a github project containing reference plot images
+    to support iris graphics tests : see :ref:`test graphics images`.
+ * If new files are required by tests or code examples, they must be added to
+   the appropriate supporting project via a suitable pull-request.
+   This new 'supporting pull request' should be referenced in the main Iris
+   pull request, and must be accepted and merged before the Iris one can be.
```
Comment (corinnebosley) [inline docs/iris/src/developers_guide/pulls.rst]: I really like your additions to this document.  Very clear and instructive.

File: docs/iris/src/developers_guide/tests.rst [hunk 4]
```diff
@@ -125,3 +127,19 @@ developers locate relevant tests. It is recommended they are named
...
 `metadata/test_pp_preservation.py`, and not named according to the
 module(s) under test.
+
+
+Graphics tests
+=================
+Certain Iris tests rely on testing plotted results.
+This is required for testing the modules :mod:`iris.plot` and
+:mod:`iris.quickplot`, but is also used for some other legacy and integration
+tests.
```
Comment (corinnebosley) [inline docs/iris/src/developers_guide/tests.rst]: There are lots of tests in this sentence.  Are there too many?

**Comment by bjlittle** [comment]
@pp-mo and @corinnebosley If https://github.com/SciTools/iris/pull/2192 is merged, then it may impact this PR.

**Comment by marqh** [comment]
#2192 is now merged and this branch has inherited a conflict

@pp-mo please may you attempt to capture these inputs and hook together @corinnebosley 's comments and update the PR

thank you

**Comment by corinnebosley** [comment]
@pp-mo I did make a branch from this branch and made some changes to the docs in your absence, but I had trouble working out how to make a PR and where to push it to and stuff, so I stopped.

**Comment by pp-mo** [comment]
> [[corinnebosley]] pp-mo I did make a branch from this branch and made some changes to the docs in your absence, but I had trouble working out how to make a PR and where to push it to and stuff, so I stopped.

Those suggestions are here : https://github.com/pp-mo/iris/pull/39

**Comment by pp-mo** [comment]
:-( Messed this up be re-using this branch
but :-) all those comments + suggestions now adopted here.

**Comment by pp-mo** [comment]
Rebased + resolved.
I've now made some response to all the previous comments + suggestions.
Please re-review !

**Comment by marqh** [comment]
this is looking pretty close to good to go to me

and remaining concerns or objections in short order please...

**Comment by marqh** [comment]
thanks @pp-mo

**Pull request merged**